base/dcp_client.go

```diff
@@ -127,26 +127,29 @@ func NewDCPClient(ID string, callback sgbucket.FeedEventCallbackFunc, options DC
 	return client, nil
 }
 
-// Start returns an error and a channel to indicate when the DCPClient is done. If Start returns an error, DCPClient.Close() needs to be called.
+// Start returns an error and a channel to indicate when the DCPClient is done.
 func (dc *DCPClient) Start() (doneChan chan error, err error) {
 	err = dc.initAgent(dc.spec)
 	if err != nil {
+		_ = dc.Close()
 		return dc.doneChannel, err
 	}
 	dc.startWorkers()
 
 	for i := uint16(0); i < dc.numVbuckets; i++ {
 		openErr := dc.openStream(i, openRetryCount)
 		if openErr != nil {
+			_ = dc.Close()
 			return dc.doneChannel, fmt.Errorf("Unable to start DCP client, error opening stream for vb %d: %w", i, openErr)
 		}
 	}
 	return dc.doneChannel, nil
 }
 
-// Close is used externally to stop the DCP client. If the client was already closed due to error, returns that error
+// Close is used externally to stop the DCP client. If the client was already closed due to error, returns that error.
 func (dc *DCPClient) Close() error {
 	dc.close()
+	<-dc.doneChannel
```

Comment:
The goroutine in dc.close() is going to write dc.getCloseError() to the channel - can we read that instead of calling again on line 153?

Also, it would be more defensive to have a select with a timeout option on this channel read, what do you think?

Reply:
The problem is that if it is already closed when it hits close(), we won't get this error, it's just a closed channel. I don't feel strongly about having a timeout, although as you say we can hit deadlocks below.

Because the dcp feed can be closed at almost any time, there's a real chance of missing the close error when it does error when you select and wait for `DCPClient.close`. The only pattern that seems OK is to wait on doneChan and not have it return errors, but check the errors after.